encrypt-config/ArgParser.cpp

```diff
@@ -0,0 +1,187 @@
+#include "utils/OptionalUtils.h"
+#include "utils/StringUtils.h"
+
+namespace org {
+namespace apache {
+namespace nifi {
+namespace minifi {
+namespace encrypt_config {
+
+const std::vector<Argument> Arguments::simple_arguments_{
+    {std::vector<std::string>{"--minifi-home", "-m"},
+     true,
+     "minifi home",
+     "Specifies the home directory used by the minifi agent"}
+};
+
+const std::vector<FlagArgument> Arguments::flag_arguments_{
+    {std::vector<std::string>{"--help", "-h"},
+     "Prints this help message"},
+    {std::vector<std::string>{"--encrypt-flow-config"},
+     "If set, the flow configuration file (as specified in minifi.properties) is also encrypted."}
+};
+
+std::string Arguments::getHelp() {
+  std::stringstream ss;
+  ss << "Usage: " << "encrypt-config";
+  for (const auto& simple_arg : simple_arguments_) {
+    ss << " ";
+    if (!simple_arg.required) {
+      ss << "[";
+    }
+    ss << utils::StringUtils::join("|", simple_arg.names)
+        << " <" << simple_arg.value_name << ">";
+    if (!simple_arg.required) {
+      ss << "]";
+    }
+  }
+  for (const auto& flag : flag_arguments_) {
+    ss << " [" << utils::StringUtils::join("|", flag.names) << "]";
+  }
+  ss << std::endl;
+  for (const auto& simple_arg : simple_arguments_) {
+    ss << "\t";
+    ss << utils::StringUtils::join("|", simple_arg.names) << " : ";
+    if (simple_arg.required) {
+      ss << "(required)";
+    } else {
+      ss << "(optional)";
+    }
+    ss << " " << simple_arg.description;
+    ss << std::endl;
+  }
+  for (const auto& flag : flag_arguments_) {
+    ss << "\t" << utils::StringUtils::join("|", flag.names) << " : "
+        << flag.description << std::endl;
+  }
+  return ss.str();
+}
+
+void Arguments::set(const std::string& key, const std::string& value) {
+  if (get(key)) {
+    std::cerr << "Key is specified more than once \"" << key << "\"" << std::endl;
+    std::cerr << getHelp();
+    std::exit(1);
+  }
+  simple_args_[key] = value;
+}
+
+void Arguments::set(const std::string& flag) {
+  if (isSet(flag)) {
+    std::cerr << "Flag is specified more than once \"" << flag << "\"" << std::endl;
+    std::cerr << getHelp();
+    std::exit(1);
+  }
+  flag_args_.insert(flag);
+}
+
+utils::optional<std::string> Arguments::get(const std::string &key) const {
+  utils::optional<Argument> opt_arg = getSimpleArg(key);
+  if (!opt_arg) {
+    return {};
+  }
+  for (const auto& name : opt_arg->names) {
+    auto it = simple_args_.find(name);
+    if (it != simple_args_.end()) {
+      return it->second;
+    }
+  }
+  return {};
+}
+
+bool Arguments::isSet(const std::string &flag) const {
+  utils::optional<FlagArgument> opt_flag = getFlag(flag);
+  if (!opt_flag) {
+    return false;
+  }
+  return std::any_of(opt_flag->names.begin(), opt_flag->names.end(), [&] (const std::string& name) {
+    return flag_args_.find(name) != flag_args_.end();
+  });
+}
+
+Arguments Arguments::parse(int argc, char* argv[]) {
```

Comment:
I wonder if we can make it throw instead of having many exit points. 

This way we could reduce the number of usage of the well-famous copy-paste pattern by only printing help and calling std::exit in the catch block.

Reply:
done